Add more smart pointers to Geolocation

abigailfox opened the pull request:
#### 71dc7942508f80337321012ad3249b0af2949fd1
<pre>
Add more smart pointers to Geolocation
<a href="https://bugs.webkit.org/show_bug.cgi?id=274489">https://bugs.webkit.org/show_bug.cgi?id=274489</a>
<a href="https://rdar.apple.com/128497803">rdar://128497803</a>

Reviewed by Ryosuke Niwa.

* Source/WebCore/Modules/geolocation/Geolocation.cpp:
(WebCore::Geolocation::page const):
(WebCore::Geolocation::resumeTimerFired):
(WebCore::Geolocation::resetAllGeolocationPermission):
(WebCore::Geolocation::stop):
(WebCore::Geolocation::lastPosition):
(WebCore::Geolocation::getCurrentPosition):
(WebCore::Geolocation::shouldBlockGeolocationRequests):
* Source/WebCore/Modules/geolocation/Geolocation.h:

Canonical link: <a href="https://commits.webkit.org/279237@main">https://commits.webkit.org/279237@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f3dc575ca9d52c01287a7c4d78fdf2f555539a3e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/52606 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/31941 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/5035 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/55908 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/3329 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/54911 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/38518 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/3030 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/42759 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/2134 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/54704 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/29628 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/45399 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/23859 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/26764 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/2683 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/1488 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/48654 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/2836 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/57476 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/27742 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/2863 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/50167 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/28967 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/45515 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/49406 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/11547 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/29881 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/28717 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/Modules/geolocation/Geolocation.cpp:
```diff
@@ -240,9 +240,9 @@ void Geolocation::resetAllGeolocationPermission()
 
     if (m_allowGeolocation == InProgress) {
-        Page* page = this->page();
+        RefPtr page = protectedPage();
```
No ned to use protectedPage here. Just do: `RefPtr page = this->page()`.

rniwa commented on Source/WebCore/Modules/geolocation/Geolocation.cpp:
```diff
@@ -270,9 +270,9 @@ void Geolocation::resetAllGeolocationPermission()
 void Geolocation::stop()
 {
-    Page* page = this->page();
+    RefPtr page = protectedPage();
```
Ditto.

rniwa commented on Source/WebCore/Modules/geolocation/Geolocation.cpp:
```diff
@@ -284,18 +284,19 @@ void Geolocation::stop()
 GeolocationPosition* Geolocation::lastPosition()
 {
-    Page* page = this->page();
+    RefPtr page = protectedPage();
```
Ditto.

rniwa commented on Source/WebCore/Modules/geolocation/Geolocation.cpp:
```diff
@@ -284,18 +284,19 @@ void Geolocation::stop()
 void Geolocation::getCurrentPosition(Ref<PositionCallback>&& successCallback, RefPtr<PositionErrorCallback>&& errorCallback, PositionOptions&& options)
 {
-    if (!document() || !document()->isFullyActive()) {
+    RefPtr document = protectedDocument();
```
Ditto.

rniwa commented on Source/WebCore/Modules/geolocation/Geolocation.cpp:
```diff
@@ -345,18 +346,19 @@ static void logError(const String& target, const bool isSecure, const bool isMix
 bool Geolocation::shouldBlockGeolocationRequests()
 {
-    if (!isPermissionsPolicyAllowedByDocumentAndAllOwners(PermissionsPolicy::Feature::Geolocation, *document(), LogPermissionsPolicyFailure::Yes))
+    RefPtr document = protectedDocument();
```
Ditto.

rniwa commented on Source/WebCore/Modules/geolocation/Geolocation.h:
```diff
@@ -65,6 +65,7 @@ class Geolocation final : public ScriptWrappable, public RefCounted<Geolocation>
 
     WEBCORE_EXPORT void resetAllGeolocationPermission();
     Document* document() const { return downcast<Document>(scriptExecutionContext()); }
+    RefPtr<Document> protectedDocument() const { return document(); }
```
These functions are useful / used when we have to call a method on it without allocating a local variable.
e.g. `protectedDocument()->foo()`.
In the case we're storing in a local variable, we don't need to use protected* functions.

rniwa commented on Source/WebCore/Modules/geolocation/Geolocation.cpp:
```diff
@@ -158,7 +158,7 @@ SecurityOrigin* Geolocation::securityOrigin() const
 Page* Geolocation::page() const
 {
-    return document() ? document()->page() : nullptr;
+    return document() ? RefPtr { document() }->page() : nullptr;
```
Instead of calling document twice. Can we do this instead?
```
RefPtr document = this->document();
return document ? document->page() : nullptr;
```

rniwa commented on Source/WebCore/Modules/geolocation/Geolocation.cpp:
```diff
@@ -345,18 +346,19 @@ static void logError(const String& target, const bool isSecure, const bool isMix
     }
 
-    logError(securityOrigin()->toString(), isSecure, hasMixedContent, document());
+    logError(RefPtr { securityOrigin() }->toString(), isSecure, hasMixedContent, document.get());
```
This will be an example of code that could benefit from having protectedSecurityOrigin() function.

rniwa commented on Source/WebCore/Modules/geolocation/Geolocation.cpp:
```diff
@@ -227,7 +227,7 @@ void Geolocation::resumeTimerFired()
 
     if (m_errorWaitingForResume) {
-        handleError(*m_errorWaitingForResume);
+        handleError(Ref { m_errorWaitingForResume.releaseNonNull() });
```
Something is really fishy here. If we're calling releaseNonNull(), we shouldn't have to manually construct Ref?

abigailfox replied:
This was an accidental amalgamation of a few things I tried, it should just be releaseNonNull(). I cleaned it up

rniwa commented on Source/WebCore/Modules/geolocation/Geolocation.cpp:
```diff
@@ -227,7 +228,7 @@ void Geolocation::resumeTimerFired()
     if (m_errorWaitingForResume) {
-        handleError(*m_errorWaitingForResume);
+        handleError(m_errorWaitingForResume.releaseNonNull());
         m_errorWaitingForResume = nullptr;
```
It seems that we can remove this line to set `m_errorWaitingForResume` to `nullptr` if we're calling `releaseNonNull`.

abigailfox committed:
Add more smart pointers to Geolocation
https://bugs.webkit.org/show_bug.cgi?id=274489
rdar://128497803

Reviewed by Ryosuke Niwa.

* Source/WebCore/Modules/geolocation/Geolocation.cpp:
(WebCore::Geolocation::page const):
(WebCore::Geolocation::resumeTimerFired):
(WebCore::Geolocation::resetAllGeolocationPermission):
(WebCore::Geolocation::stop):
(WebCore::Geolocation::lastPosition):
(WebCore::Geolocation::getCurrentPosition):
(WebCore::Geolocation::shouldBlockGeolocationRequests):
* Source/WebCore/Modules/geolocation/Geolocation.h:

Canonical link: https://commits.webkit.org/279237@main
